all: make it easier to migrate off of gopkg.in/check.v1

lmb opened the pull request:
**Merge after the community meeting on 2023-05-24. Since this PR touches a lot of files it won't require all CODEOWNERs to sign off before merging.** Feel free to provide feedback if you've been tagged for review, but it's not a requirement.

testutils: remove PrivilegedCheck

    Now that check.C embeds testing.T we don't need separate testutils anymore.
    This shows how we can introduce helpers that are available from both testing
    and check style tests.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

docs: explain how to migrate from gocheck

    Remove outdated documentation relating to gocheck and explain how existing
    tests can be translated from gocheck to testing.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

golangci: prohibit use of gopkg.in/check.v1

    Configure golangci-lint so that we don't add new code which uses the
    deprecated package.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

all: replace gopkg.in/check.v1 with shim

    There are a lot of tests in the codebase which use gopkg.in/check.v1, which
    hasn't seen an update in the last three years. A while back it was decided
    that we'd migrate away from the package and instead use plain old stdlib
    testing.

    In practice things aren't so simple. We have so many tests that migrating
    them in one fell swoop is impossible. At the same time, the API exposed by
    check is entirely incompatible with testing. It's not possible to write a
    helper function which works for both check and testing. So far we've opted
    to provide two separate helpers, for example PrivilegedTest and
    PrivilegedCheck. This causes a lot of duplication and discourages sharing
    helpers. It also makes refactoring code from check to testing harder since
    the natural way to express the SetUp and TearDown behaviour of check is such
    a helper function.

    Additionally, check is a burden on anyone trying to run the tests: it has
    completely separate flags which control what tests are being run. The output
    also doesn't match what other go tooling expects, both for tests and
    benchmarks.

    To fix both of these, replace gopkg.in/check.v1 with a fork that removes
    most command line flags and gets rid of the custom runner. Instead, suites
    and tests are run as subtests. The other API should be mostly compatible.
    Most importantly check.C now embeds testing.T. As a result, check.C
    satisfies testing.TB, which allows us to access helpers written for the
    testing package from within unmigrated tests.

    Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

```release-note
Make it easier to migrate off of gopkg.in/check.v1
```

lmb committed:
testutils: remove PrivilegedCheck

Now that check.C embeds testing.T we don't need separate testutils
anymore. This shows how we can introduce helpers that are available
from both testing and check style tests.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
docs: explain how to migrate from gocheck

Remove outdated documentation relating to gocheck and explain how
existing tests can be translated from gocheck to testing.

Co-authored-by: ZSC <zacharysarah@users.noreply.github.com>
Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

lmb committed:
all: replace gopkg.in/check.v1 with shim

There are a lot of tests in the codebase which use gopkg.in/check.v1,
which hasn't seen an update in the last three years. A while back it
was decided that we'd migrate away from the package and instead use
plain old stdlib testing.

In practice things aren't so simple. We have so many tests that
migrating them in one fell swoop is impossible. At the same time,
the API exposed by check is entirely incompatible with testing.
It's not possible to write a helper function which works for both
check and testing. So far we've opted to provide two separate
helpers, for example PrivilegedTest and PrivilegedCheck. This causes
a lot of duplication and discourages sharing helpers. It also
makes refactoring code from check to testing harder since the natural
way to express the SetUp and TearDown behaviour of check is
such a helper function.

Additionally, check is a burden on anyone trying to run the tests:
it has completely separate flags which control what tests are
being run. The output also doesn't match what other go tooling
expects, both for tests and benchmarks.

To fix both of these, replace gopkg.in/check.v1 with a fork that
removes most command line flags and gets rid of the custom runner.
Instead, suites and tests are run as subtests. The other API should
be mostly compatible. Most importantly check.C now embeds testing.T.
As a result, check.C satisfies testing.TB, which allows us to
access helpers written for the testing package from within unmigrated
tests.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

ti-mo commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +108,120 @@ run this in a terminal next to your editor:
+        assert.Equal(c, "42", "42")
+    }
+
+As you can see we didn't get round to converting all tests yet, but that is OK.
```
I really wish we'd move away from the 'suite' concept, at least for new code, as it tightly couples the environments of multiple tests. It would be useful to reaffirm this in an additional short paragraph. Currently, this could be interpreted as "adding tests leaning on checkmate is OK". If you're working on extending test coverage, taking some extra time to lift a suite out of check.C would be much appreciated.

lmb replied:
Agreed and this isn't meant to encourage using suite. Reworded a bit and shuffled some of the doc around.

ti-mo reviewed (approved):
![It's a yes from me](https://media.tenor.com/66SLqKtBBgIAAAAC/im-gonna-say-yes-david-walliams.gif)

tklauser commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +108,120 @@ run this in a terminal next to your editor:
+
+    func Test(t *testing.T) { TestingT(t) }
+
+    func setupHelper(tb testing.TB) {
```
nit: in my experience it's good practice to mark test helpers using [`TB.Helper`](https://pkg.go.dev/testing#T.Helper) so that file and line information of the caller will be reported instead of the ones of the helper.

```suggestion
    func setupHelper(tb testing.TB) {
        tb.Helper()
```

lmb committed:
golangci: prohibit use of gopkg.in/check.v1

Configure golangci-lint so that we don't add new code which uses
the deprecated package.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

bimmlerd reviewed (approved):
Awesome :1st_place_medal:

giorio94 commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,121 @@ run this in a terminal next to your editor:
+Since ``C`` now implements ``testing.TB`` we can call ``assert.Equal`` from
+``TestTheRealAnswer()`` without problems.
+
+.. _assert: https://github.com/stretchr/testify/assert
```
This link leads to a 404. Should it be https://pkg.go.dev/github.com/stretchr/testify/assert?

giorio94 commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,121 @@ run this in a terminal next to your editor:
+
+The Cilium codebase has historically used ``gopkg.in/check.v1`` to write unit and
+integration tests. Most of the functionality offered by that package is now
+present in `testing`_. What ``gocheck`` calls checkers are now available in `assert`_:
```
More a question than a comment. Is it there any guideline about [*assert*](https://pkg.go.dev/github.com/stretchr/testify/assert) vs [*require*](https://pkg.go.dev/github.com/stretchr/testify/require)? It seems that both are used in the repo currently, but this doc mentions *assert* only.

chancez replied:
You assert when the following assertions do not require the assertions before to hold true. You require when you do. So if you're going to index a slice, you should require.Len to be non-zero before you do the indexing, otherwise your test might panic. Usually it's most useful for checking errors since your other return value will probably be nil.

giorio94 replied:
Thanks for the clarification, makes definitely sense to me. I personally tend to mostly use *require* to be on the safe side, given that in many cases an error would cause a cascade of failures. But since here I saw that only *assert* was mentioned, I started wondering whether there was any different guideline...

chancez replied:
Yeah, the interfaces are the same, use whatever makes the most sense for the situation you're in.

lmb replied:
I'm probably in the require camp myself, didn't realise that assert didn't `Fatal` the test. I'll add a little blurb.

lmb replied:
Turns out that `C.Assert` behaves like `require.*`, so I updated the doc accordingly.

giorio94 reviewed (approved):
Thanks!

tommyp1ckles reviewed (approved):
This is a really good change - Azure LGTM

christarazi reviewed (approved):
:broom:

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -9,9 +9,24 @@
-in combination with `gocheck <http://labix.org/gocheck>`__ for richer testing
-functionality.
+Cilium uses the standard `go test <https://golang.org/pkg/testing/>`__ framework.
+All new tests `must use testing`_.
```
Edits for clarity, specific is better than general. This also affects link formatting 👇🏻 
```suggestion
All new tests must use `the standard test framework`_.
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -9,9 +9,24 @@
+
+Historically, `gocheck <https://gopkg.in/check.v1>`__ has also been used and
+there are still many tests written using ``gocheck`` today, which has some
+unfortunate consequences.
```
Edits for clarity
```suggestion
Historically, `gocheck <https://gopkg.in/check.v1>`__ has also been used. 
Many tests are still written using ``gocheck``, which creates some
unfortunate consequences:
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -9,9 +9,24 @@
+1. Adding a test helper requires writing two versions: one for ``testing.T``
+   and one for ``check.C``.
+
+2. Executing those tests requires using non-standard flags.
```
Specific is better than general
```suggestion
2. Executing ``gocheck`` tests requires non-standard flags.
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -9,9 +9,24 @@
+For this reason ``gopkg.in/check.v1`` has been switched to a fork called
+``github.com/cilium/checkmate``, which removes most of the problematic bits.
+Please help migrating away from ``gocheck`` by following the procedure outlined
+in :ref:`migrate-gocheck`.
```
Edits for clarity, use active voice, use plain, clear language.
```suggestion
For this reason ``gopkg.in/check.v1`` has switched to a fork called
``github.com/cilium/checkmate``, which removes most of the problematic bits.
You can migrate away from ``gocheck`` by following the procedure outlined
in :ref:`migrate-gocheck`.
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -9,9 +9,24 @@
+Please help migrating away from ``gocheck`` by following the procedure outlined
+in :ref:`migrate-gocheck`.
+
+.. _must use testing: https://github.com/cilium/cilium/issues/16860
```
Updated link format ☝🏻 
```suggestion
.. _the standard test framework: https://github.com/cilium/cilium/issues/16860
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,125 @@ run this in a terminal next to your editor:
+The Cilium codebase has historically used ``gopkg.in/check.v1`` to write unit and
+integration tests. Most of the functionality offered by that package is now
+present in `testing`_. What ``gocheck`` calls checkers are now available in
+``testify``, either in `assert`_ or `require`_ form.
```
Clarity
```suggestion
``testify``, in either `assert`_ or `require`_ form.
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,125 @@ run this in a terminal next to your editor:
+
+* ``checker.Equals`` and ``checker.DeepEquals``: ``require.Equal``
+
+It is best to replace ``check.C.Assert`` with ``require`` calls instead of ``assert``, since the latter doesn't stop test execution if the assertion fails.
```
s/since/because to clarify the causative relationship
```suggestion
It's best to replace ``check.C.Assert`` with ``require`` calls instead of ``assert`` because the latter doesn't stop test execution if the assertion fails.
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,125 @@ run this in a terminal next to your editor:
+
+It is best to replace ``check.C.Assert`` with ``require`` calls instead of ``assert``, since the latter doesn't stop test execution if the assertion fails.
+
+``gopkg.in/check.v1`` has been replaced with ``github.com/cilium/checkmate`` which
```
Comma nit
```suggestion
``gopkg.in/check.v1`` has been replaced with ``github.com/cilium/checkmate``, which
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,125 @@ run this in a terminal next to your editor:
+and can be passed to test helpers that take ``testing.TB``.
+
+The end goal is to remove all uses of ``checkmate`` from the codebase.
+To convert a ``gocheck`` test you can use the following approach:
```
Style
```suggestion
To convert a ``gocheck`` test, use the following approach:
```

zacharysarah commented on Documentation/contributing/testing/unit.rst:
```diff
@@ -123,3 +119,125 @@ run this in a terminal next to your editor:
+        c.Assert("42", Equals, "42")
+    }
+
+After applying the above rules we might end up with something like this:
```
s/above/previous, address the reader as "you"
```suggestion
After applying the previous rules you should end up with something like this:
```

zacharysarah reviewed (approved):
@lmb 👋🏻 Some style nits, otherwise docs LGTM. Approving with the understanding that changes must be made prior to merge.
